runtime/client_http_request.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zanzibar
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/uber-go/zap"
+)
+
+// ClientHTTPRequest is the struct for making outbound
+// requests using a client.
+type ClientHTTPRequest struct {
+	started     bool
+	startTime   time.Time
+	client      *HTTPClient
+	httpRequest *http.Request
+
+	ClientName string
+	MethodName string
+	Logger     zap.Logger
+}
+
+// NewClientHTTPRequest allocates a ClientHTTPRequest
+func NewClientHTTPRequest(
+	clientName string, methodName string,
+	client *HTTPClient,
+) *ClientHTTPRequest {
+	req := &ClientHTTPRequest{
+		Logger: client.Logger,
+		client: client,
+	}
+
+	req.start(clientName, methodName)
```

Comment:
There are many timers that are interesting to us.

For example "when an httpclientrequest is started, including serialize/deserializer" and "when a http request is send until a response is received, just network IO".

TChannel adds the start timer at request creation time ( https://github.com/uber/tchannel-go/blob/3abb4c025c1663b383452339a22d918cf9d0be0b/outbound.go#L119 ).